owners/src/owners.js

```diff
@@ -18,6 +18,129 @@ const path = require('path');
 const yaml = require('yamljs');
 
 
+/**
+ * A tree of ownership keyed by directory.
+ */
+class OwnersTree {
+  /**
+   * Constructor.
+   *
+   * @param {!string=} dirPath relative the directory containing owners file.
+   * @param {OwnersTree=} parent parent tree of the node (null for root)
+   */
+  constructor(dirPath, parent) {
+    this.parent = parent || null;
+    this.dirPath = dirPath || '.';
+
+    this.isRoot = !parent;
+
+
+    this.rules = [];
+    this.children = {};
+  }
+
+  /**
+   * Add a rule to the tree.
+   *
+   * Adds to the current node if paths match; otherwise, adds it to the node
+   * keyed by the next directory name.
+   *
+   * @param {!OwnersRule} rule rule to add.
+   */
+  addRule(rule) {
+    if (this.depth > 5) {
```

Comment:
Why 5? What if it's deeper than 5?

Reply:
See this is why code reviews are helpful :sweat_smile: This was in here while I was fixing an infinite-recursion bug, and I neglected to add a `TODO`